File: src/evidence/privacy.ts
/**
 * 隐私过滤：本地优先，并严格执行证据级授权位。
 *
 * filterReadableByTier：从【提供给模型的输入】中排除当前模型 tier 无权读取的证据。
 *   tier='cloud'：只留 allowCloudRead=true（进入 MemoWeft 内建云写模型 prompt）。
 *   tier='local'：只留 allowLocalRead=true（进入本地写模型 prompt，可读 observed 默认不进入云写 prompt 的那些）。
 * 写路径各处取证据喂 LLM 前共用这道关，按 deps.llm 的 tier（缺省 'cloud'，最保守）决定筛哪个授权位。
 *
 * ⚠️ 两维正交，别混：本关只管【内建写模型 prompt 的证据选择】（云/本地 = allowCloudRead / allowLocalRead）。是否能据此推画像
 *   是另一维 allowInference——由画像/推理路径（distill / consolidate / attribute）在本关之外另行
 *   `.filter((e) => e.allowInference)`（三处口径一致，防 inference=false 的证据经 event summary 间接渗进画像）。
 */
import type { ModelTier } from '../llm/client.ts';

/** 按模型 tier 保留"该 tier 允许读"的项；其余原样顺序保留。缺省 tier = 'cloud'（最保守：不误把敏感证据当本地放行）。 */
export function filterReadableByTier<
  T extends { allowCloudRead: boolean; allowLocalRead: boolean },
>(items: T[], tier: ModelTier = 'cloud'): T[] {
  return tier === 'local'
    ? items.filter((e) => e.allowLocalRead)
    : items.filter((e) => e.allowCloudRead);
}

/**
 * @deprecated 用 `filterReadableByTier(items, tier)`（等价于 tier='cloud'）。
 * 库内已全切 filterReadableByTier；本别名仅保兼容（privacy.test 仍直接测它、外部旧调用方兜底）。
 */
export function filterCloudReadable<T extends { allowCloudRead: boolean }>(items: T[]): T[] {
  return items.filter((e) => e.allowCloudRead);
}

File: py/src/memoweft/privacy.py
"""与 TypeScript evidence privacy 契约保持一致的隐私读取边界。

filter_readable_by_tier:从提供给当前模型的材料中排除该 tier 无权读取的证据。
  cloud → 留 allow_cloud_read;local → 留 allow_local_read。缺省 cloud(最保守)。保序过滤。
⚠️ 只管读取权;是否可推画像是另一维 allow_inference,由 distill/consolidate/attribute 在本关之外另筛。
"""
from __future__ import annotations

from typing import Protocol, TypeVar

from .types import ModelTier


class _Readable(Protocol):
    # 用 @property(只读)—— frozen dataclass(Evidence)的只读字段才满足 Protocol 约束。
    @property
    def allow_cloud_read(self) -> bool: ...
    @property
    def allow_local_read(self) -> bool: ...


_T = TypeVar("_T", bound=_Readable)


def filter_readable_by_tier(items: list[_T], tier: ModelTier = "cloud") -> list[_T]:
    """按 tier 保留允许读取的项并维持顺序；未指定 tier 时使用保守的 cloud 默认值。"""
    if tier == "local":
        return [e for e in items if e.allow_local_read]
    return [e for e in items if e.allow_cloud_read]


class _CloudReadable(Protocol):
    @property
    def allow_cloud_read(self) -> bool: ...


_C = TypeVar("_C", bound=_CloudReadable)


def filter_cloud_readable(items: list[_C]) -> list[_C]:
    """保留用于兼容的已弃用别名，等价于 tier='cloud'。"""
    return [e for e in items if e.allow_cloud_read]
